src/vm/docs/mod.rs

```diff
@@ -538,6 +540,96 @@ The `header-hash`, `burnchain-header-hash`, and `vrf-seed` properties return a 3
 "
 };
 
+const DEFINE_PUBLIC_API: SpecialAPI = SpecialAPI {
+    name: "define-public",
+    input_type: "MethodSignature, MethodBody",
+    output_type: "",
+    signature: "(define-public (function-name (arg-name-0 arg-type-0) (arg-name-1 arg-type-1) ...) function-body)",
+    description: "`define-public` is used to define a _public_ function and transaction for a smart contract. Public
+functions are callable from other smart contracts, and may be invoked directly by users by submitting a transaction
+to the Stacks blockchain.
+
+Like other kinds of definition statements, `define-public` may only be used at the top level of a smart contract
+definition (i.e., you cannot put a define statement in the middle of a function body).
+
+Public functions _must_ return a ResponseType (using either `ok` or `err`). Any datamap modifications performed by
+a public function will be aborted if the function returns an `err` type. Public functions may be invoked by other
+contracts via `contract-call!`.",
```

Comment:
Even if the function invoking `contract-call!` handles the `(err` result, no changes will be materialized?

Reply:
Correct -- anytime a function returns `err`, any of the modifications it (or functions it invoked) made are aborted. If you want a function to return different kinds of values while still making database changes, you would have to use the `ok` type.

Reply:
The _caller_, if it handles the `err` result can still materialize its own changes, though.

Reply:
That's what I thought.  Thanks for clarifying!